measure-android/measure/src/main/java/sh/measure/android/screenshot/ScreenshotCollector.kt

```diff
@@ -0,0 +1,207 @@
+package sh.measure.android.screenshot
+
+import android.app.Activity
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.Rect
+import android.graphics.RectF
+import android.os.Build
+import android.os.Handler
+import android.os.HandlerThread
+import android.view.PixelCopy
+import android.view.View
+import android.view.Window
+import androidx.annotation.RequiresApi
+import sh.measure.android.Config
+import sh.measure.android.isMainThread
+import sh.measure.android.logger.LogLevel
+import sh.measure.android.logger.Logger
+import sh.measure.android.tracing.InternalTrace
+import sh.measure.android.utils.LowMemoryCheck
+import sh.measure.android.utils.ResumedActivityProvider
+import java.io.ByteArrayOutputStream
+import kotlin.LazyThreadSafetyMode.NONE
+
+internal class Screenshot(
+    val data: ByteArray,
+    val extension: String,
+)
+
+internal interface ScreenshotCollector {
+    /**
+     * Attempts to take a screenshot of the visible activity, if any.
+     *
+     * @return The screenshot as a byte array, or null if the screenshot could not be taken.
+     */
+    fun takeScreenshot(): Screenshot?
+}
+
+/**
+ * Captures a screenshot of the currently resumed activity, given the system memory is not running
+ * low. The screenshot is captured using the PixelCopy API on Android O and above, and using
+ * the Canvas API on older versions. The returned screenshot is compressed webp on supported
+ * versions with a fallback to JPEG.
+ *
+ * The screenshot is masked to hide sensitive text and can also be configured to mask all text.
+ */
+internal class ScreenshotCollectorImpl(
+    private val logger: Logger,
+    private val resumedActivityProvider: ResumedActivityProvider,
+    private val lowMemoryCheck: LowMemoryCheck,
+    private val config: Config,
+) : ScreenshotCollector {
+    private val maskPaint by lazy(NONE) {
+        Paint().apply {
+            color = Color.parseColor(config.screenshotMaskHexColor)
+            style = Paint.Style.FILL
+        }
+    }
+    private val maskRadius = config.screenshotMaskRadius
+    private val webpScreenshotCompression = config.screenshotWebpQuality
+    private val jpegScreenshotCompression = config.screenshotWebpQuality
+
+    override fun takeScreenshot(): Screenshot? {
+        if (lowMemoryCheck.isLowMemory()) {
+            logger.log(
+                LogLevel.Debug,
+                "Unable to take screenshot, system has low memory.",
+            )
+            return null
+        }
+
+        return resumedActivityProvider.getResumedActivity()?.let {
```

Comment:
@abhaysood , what happens if the activity crashes during onCreate or onStart? ( we will end-up capturing the wrong screenshot )

Reply:
The first frame of an activity is typically drawn after onResume. If an activity crashes before onResume it would likely not have rendered its content.

Reply:
We capture screenshots at the time of crash as of now. We intend to provide the developer context about where the user was at the time of crash.